leader/poller.go

```diff
@@ -0,0 +1,75 @@
+package leader
+
+import (
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+type poller struct {
+	leaderChan   chan Leadership
+	pollInterval time.Duration
+	tick         <-chan time.Time
+	stop         chan struct{}
+	pollFunc     CheckLeaderFunc
+}
+
+// NewPoller returns a detector implementation given the poll interval and function that polls
+func NewPoller(pollInterval time.Duration, f CheckLeaderFunc) Detector {
+	return &poller{
+		pollInterval: pollInterval,
+		tick:         time.Tick(pollInterval),
+		pollFunc:     f,
+	}
+}
+
+// Start implements Detect.Start
+func (l *poller) Start() (<-chan Leadership, error) {
+	if l.leaderChan != nil {
+		return l.leaderChan, nil
+	}
+
+	l.leaderChan = make(chan Leadership)
+	l.stop = make(chan struct{})
+
+	go l.poll()
+	return l.leaderChan, nil
+}
+
+// Stop implements Detect.Stop
+func (l *poller) Stop() {
+	if l.stop != nil {
+		close(l.stop)
+	}
+}
+
+func (l *poller) poll() {
+	for {
+		select {
+
+		case <-l.tick:
+
+			isLeader, err := l.pollFunc()
+			event := Leadership{}
+			if err != nil {
+				event.Status = StatusUnknown
+				event.Error = err
+			} else {
+				if isLeader {
+					event.Status = StatusLeader
+				} else {
+					event.Status = StatusNotLeader
+				}
+			}
+
+			l.leaderChan <- event
```

Comment:
How would you feel about making events edge-triggered instead of 'pings'?  In addition to making the system more chatty (and natural to build with an external event-based system), it will make it more apparent where to add support for something like timeouts for unknown leadership.

This should just be a matter of tracking the last event sent and suppressing if equal.